### PR TITLE
feat/address-service/doma-5952/bulk search for billing

### PR DESCRIPTION
**Pull request by AleX83Xpert**

### CODE REVIEW

**Review by dkoviazin — APPROVED** [review]

File: apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js [hunk 1]
```diff
@@ -42,27 +47,104 @@ class SearchKeystoneApp {
...
+                if (!!ut && !!un) {
+                    unitType = ut
+                    unitName = un
+                }
+            }
+
+            const pluginParams = { searchContext, keystoneContext, req, helpers }
+
+            const plugins = this.plugins.filter((plugin) => plugin.isEnabled(searchString, pluginParams))
+            if (plugins.length === 0) {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js:88]: if (isEmpty(plugins))

File: apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js [hunk 2]
```diff
@@ -42,27 +47,104 @@ class SearchKeystoneApp {
...
+                return { err: SEARCH_ERROR_NO_PLUGINS }
+            }
+
+            let searchResult = null
+            try {
+                for (const plugin of plugins) {
+                    // Return the first not empty plugin's result
+                    // So, the plugins order is mandatory for performance
+                    if (searchResult) {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js:97]: if (searchResult) break

File: apps/address-service/domains/common/utils/services/search/plugins/SearchBySource.spec.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    let sender
+
+    setFakeClientMode(index)
+
+    beforeAll(async () => {
+        context = await keystone.createContext({ skipAccessControl: true })
+        adminClient = await makeLoggedInAdminClient()
+        sender = { dv: 1, fingerprint: faker.random.alphaNumeric(8) }
+    })
+    test('helpers using correctly', async () => {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/plugins/SearchBySource.spec.js:35]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.js [hunk 4]
```diff
@@ -84,4 +94,48 @@ async function createReturnObject ({
     return ret
 }
 
-module.exports = { createOrUpdateAddressWithSource, createReturnObject }
+function sortObject (obj) {
+    if (typeof obj !== 'object' || obj === null) {
+        return obj
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.js:99]: when you use single return in condition, no curly braces required, you can write it in one line of code

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.js [hunk 5]
```diff
@@ -84,4 +94,48 @@ async function createReturnObject ({
...
+    }
+
+    if (Array.isArray(obj)) {
+        return obj.map(sortObject)
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.js:103]: same

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.js [hunk 6]
```diff
@@ -84,4 +94,48 @@ async function createReturnObject ({
...
+        result[key] = sortObject(obj[key])
+    })
+    return result
+}
+
+function hashJSON (obj) {
+    const sortedObj = sortObject(obj)
+    const jsonStr = JSON.stringify(sortedObj)
+    const hash = createHash('md5')
+    return hash.update(jsonStr).digest('hex')
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.js:118]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.js [hunk 7]
```diff
@@ -84,4 +94,48 @@ async function createReturnObject ({
...
+    if (Array.isArray(obj)) {
+        return obj.map(sortObject)
+    }
+
+    const sortedKeys = Object.keys(obj).sort()
+    const result = {}
+    sortedKeys.forEach((key) => {
+        result[key] = sortObject(obj[key])
+    })
+    return result
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.js:111]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.js [hunk 8]
```diff
@@ -84,4 +94,48 @@ async function createReturnObject ({
...
+        return obj
+    }
+
+    if (Array.isArray(obj)) {
+        return obj.map(sortObject)
+    }
+
+    const sortedKeys = Object.keys(obj).sort()
+    const result = {}
+    sortedKeys.forEach((key) => {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.js:108]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.js [hunk 9]
```diff
@@ -84,4 +94,48 @@ async function createReturnObject ({
...
+}
+
+/**
+ * @param {string} address
+ * @param {?SuggestionHelpersType} [helpers]
+ * @returns {string}
+ */
+function mergeAddressAndHelpers (address, helpers) {
+    if (!isObject(helpers) || isEmpty(helpers)) {
+        return address
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.js:128]: same here

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            const hash = hashJSON(obj)
+            expect(hash).toEqual('608de49a4600dbb5b173492759792e4a')
+        })
+
+        test('should return the same hash for objects with the same properties in different order', () => {
+            const obj1 = { a: 1, b: 2 }
+            const obj2 = { b: 2, a: 1 }
+            const hash1 = hashJSON(obj1)
+            const hash2 = hashJSON(obj2)
+            expect(hash1).toEqual(hash2)
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js:18]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        })
+
+        test('should return a different hash for objects with different properties', () => {
+            const obj1 = { a: 1, b: 2 }
+            const obj2 = { c: 3, d: 4 }
+            const hash1 = hashJSON(obj1)
+            const hash2 = hashJSON(obj2)
+            expect(hash1).not.toEqual(hash2)
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js:26]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        })
+
+        test('should return a different hash for objects with nested objects with different properties', () => {
+            const obj1 = { a: 1, b: { c: 2, d: 3 } }
+            const obj2 = { a: 1, b: { c: 3, d: 2 } }
+            const hash1 = hashJSON(obj1)
+            const hash2 = hashJSON(obj2)
+            expect(hash1).not.toEqual(hash2)
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js:34]: empty line

File: apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        })
+
+        test('should return the same hash for objects with nested objects with the same properties in different order', () => {
+            const obj1 = { a: 1, b: { c: 2, d: 3 } }
+            const obj2 = { a: 1, b: { d: 3, c: 2 } }
+            const hash1 = hashJSON(obj1)
+            const hash2 = hashJSON(obj2)
+            expect(hash1).toEqual(hash2)
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/search/searchServiceUtils.spec.js:42]: empty line

File: apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js [hunk 14]
```diff
@@ -81,53 +163,97 @@ class SearchKeystoneApp {
...
+             * @returns {Promise<void>}
+             */
+            async (req, res, next) => {
+                /** @type {String[]} */
+                const items = get(req, ['body', 'items'])
 
-                if (!searchResult) {
-                    // Nothing found
-                    res.send(404)
+                if (!items) {
```
Comment (Alllex202) [inline apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js:219]: Maybe check that an array is passed?

**Review by Alllex202 — APPROVED** [review]

File: apps/address-service/domains/common/utils/parseAddressesFromString.spec.js [hunk 15]
```diff
@@ -24,7 +24,7 @@ describe('Parsing unit names with unit types and house address from address stri
     for (const rawData of ADDRESS_USE_CASES) {
         const [ rawInput, house, unitType, unit ] = rawData
         test(`"${rawInput}" to be: "${house} = ${unitType} = ${unit}" `, () => {
-            const [{ result: { address, unit: parsedUnit, unitType: parsedUnitType } }] = parseAddressesFromString([ rawInput ])
+            const [{ result: { address, unitName: parsedUnit, unitType: parsedUnitType } }] = parseAddressesFromString([ rawInput ])
```
Comment (pahaz) [inline apps/address-service/domains/common/utils/parseAddressesFromString.spec.js:27]: parsedUnitName

**Review by pahaz — APPROVED** [review]

File: apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js (hunk 14 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/search/SearchKeystoneApp.js:219]: Good suggestion, thanks!
Done.

File: apps/address-service/domains/common/utils/parseAddressesFromString.spec.js (hunk 15 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/parseAddressesFromString.spec.js:27]: Fixed

**Pull request merged**